readme.md
<!--
  Created at: 08/31/2021 21:47:37 Tuesday
  Modified at: 09/24/2021 08:05:46 PM Friday

        Copyright (C) 2021 Thiago Navarro
  See file "license" for details about copyright
-->

# oscommerce2woocommerce

OsCommerce database to WooCommerce compatible

## WARNING

This project was abandoned because [this wp module](https://github.com/alexwing/woocommerce-oscommerce-sync) solves well

## Features

- [ ] Convertion
  - [x] customers
  - [x] address_book
  - [ ] orders
- [ ] Saving
  - [ ] customers
  - [ ] address_book
  - [ ] orders

## Todo

- [ ] Add more data convertion
- [ ] Test older Nim versions

## Docs links

The following urls is of Woocommerce REST API documentation

## Customer

- [Customer properties](https://woocommerce.github.io/woocommerce-rest-api-docs/?shell#customer-properties)
- [Create a customer](https://woocommerce.github.io/woocommerce-rest-api-docs/?shell#create-a-customer)

## Orders

- [Order properties](https://woocommerce.github.io/woocommerce-rest-api-docs/?javascript#order-properties)
- [Create an order](https://woocommerce.github.io/woocommerce-rest-api-docs/?javascript#create-an-order)

## Order Notes

- [Order note properties](https://woocommerce.github.io/woocommerce-rest-api-docs/?javascript#order-note-properties)
- [Create an order note](https://woocommerce.github.io/woocommerce-rest-api-docs/?javascript#create-an-order-note)

## License

GPLv3


changelog.md
<!--
  Created at: 09/22/2021 17:00:50 Wednesday
  Modified at: 09/22/2021 05:01:57 PM Wednesday

        Copyright (C) 2021 Thiago Navarro
  See file "license" for details about copyright
-->

# Changelog

## Version 0.1.1 (09/22/2021 17:01:57 Wednesday)

- fixed catch exception in main module
- added test dummy data
- added `out.json` to `.gitignore`


tests/test1.nim
# This is just an example to get you started. You may wish to put all of your
# tests into a single file, or separate them into multiple `test1`, `test2`
# etc. files (better names are recommended, just make sure the name starts with
# the letter 't').
#
# To run these tests, simply execute `nimble test`.

import unittest

import oscommerce2woocommerce
test "can add":
  check add(5, 5) == 10


tests/testCustomers.nim
#[
  Created at: 08/31/2021 21:56:51 Tuesday
  Modified at: 09/28/2021 06:34:04 PM Tuesday

        Copyright (C) 2021 Thiago Navarro
  See file "license" for details about copyright
]#

import std/[unittest, strformat]
import oscommerce2woocommerce

const osCommerceUser = OsCommerceUser(
  customers_id: 10,
  customers_gender: "m",
  customers_firstname: "John",
  customers_lastname: "Doe",
  customers_dob: "1950-1-1 00:00:00",
  customers_email_address: "example@example.com",
  customers_default_address_id: 1,
  customers_telephone: "123456789",
  customers_fax: "",
  customers_password: "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa:aa",
  customers_newsletter: "1",
  customers_activation_code: "NULL",
  customers_active_status: "1",
  customers_notes: "Hello World!",
  address: OsCommerceUserAddress(
    address_book_id: 1,
    customers_id: 1,
    entry_gender: "m",
    entry_company: "Company of Johns",
    entry_firstname: "John",
    entry_lastname: "Doe",
    entry_street_address: "Street of Johns",
    entry_suburb: "suburb",
    entry_postcode: "12345678",
    entry_city: "City of Johns",
    entry_state: "State of Johns",
    entry_country_id: 30,
    entry_zone_id: 0,
    entry_street_address2: "",
    entry_street_address3: ""
  )
)

suite "customers":
  let wooCommerceUser = osCommerceUser.toWoo()
  test "First name":
    check wooCommerceUser.first_name == osCommerceUser.customers_firstname
  test "Last name":
    check wooCommerceUser.last_name == osCommerceUser.customers_lastname
  test "Email":
    check wooCommerceUser.email == osCommerceUser.customers_email_address
  test "Username":
    check wooCommerceUser.username == "thiago"
  test "Billing first name":
    check wooCommerceUser.billing.first_name == osCommerceUser.customers_firstname
  test "Billing last name":
    check wooCommerceUser.billing.last_name == osCommerceUser.customers_lastname
  test "Billing email":
    check wooCommerceUser.billing.email == osCommerceUser.customers_email_address
  test "Billing phone":
    check wooCommerceUser.billing.phone == osCommerceUser.customers_telephone
  test "Billing company":
    check wooCommerceUser.billing.company == ""
  test "Billing address_1":
    check wooCommerceUser.billing.address_1 == osCommerceUser.address.entry_street_address
  test "Billing address_2":
    check wooCommerceUser.billing.address_2 == ""
  test "Billing city":
    check wooCommerceUser.billing.city == osCommerceUser.address.entry_city
  test "Billing state":
    check wooCommerceUser.billing.state == ""
  test "Billing postcode":
    check wooCommerceUser.billing.postcode == osCommerceUser.address.entry_postcode
  test "Billing country":
    check wooCommerceUser.billing.country == "BR"
  test "Shipping first name":
    check wooCommerceUser.shipping.first_name == osCommerceUser.customers_firstname
  test "Shipping last name":
    check wooCommerceUser.shipping.last_name == osCommerceUser.customers_lastname
  test "Shipping company":
    check wooCommerceUser.shipping.company == ""
  test "Shipping address_1":
    check wooCommerceUser.shipping.address_1 == osCommerceUser.address.entry_street_address
  test "Shipping address_2":
    check wooCommerceUser.shipping.address_2 == ""
  test "Shipping city":
    check wooCommerceUser.shipping.city == osCommerceUser.address.entry_city
  test "Shipping state":
    check wooCommerceUser.shipping.state == ""
  test "Shipping postcode":
    check wooCommerceUser.shipping.postcode == osCommerceUser.address.entry_postcode
  test "Shipping country":
    check wooCommerceUser.shipping.country == "BR"


a.sh
#!/bin/bash
# curl -X POST http://127.0.0.1/u3/apache/5.8/wp-json/wc/v3/customers \
curl -X POST http://127.0.0.1:3001/wp-json/wc/v3/customers \
    -u ck_f669a0590f054d89fee6a36dfe4cfc2ed06f48d5:cs_1e33fcd228feb3070adc9f17080822b637787464 \
    -H "Content-Type: application/json" \
    -d '{
  "email": "john.doe@example.com",
  "first_name": "John",
  "last_name": "Doe",
  "username": "john.doe",
  "billing": {
    "first_name": "John",
    "last_name": "Doe",
    "company": "",
    "address_1": "969 Market",
    "address_2": "",
    "city": "San Francisco",
    "state": "CA",
    "postcode": "94103",
    "country": "US",
    "email": "john.doe@example.com",
    "phone": "(555) 555-5555"
  },
  "shipping": {
    "first_name": "John",
    "last_name": "Doe",
    "company": "",
    "address_1": "969 Market",
    "address_2": "",
    "city": "San Francisco",
    "state": "CA",
    "postcode": "94103",
    "country": "US"
  }
}'


src/oscommerce2woocommerce/base.nim
#[
  Created at: 08/31/2021 23:04:55 Tuesday
  Modified at: 09/01/2021 08:45:06 PM Wednesday

        Copyright (C) 2021 Thiago Navarro
  See file "license" for details about copyright
]#

{.experimental: "codeReordering".}

type
  WooData* = object
    key*: WooKey
    value*: string
    discarded*: bool
  WooKey* = tuple
    database, column: string

  OsCommerceUser* = object
    customers_id*, customers_default_address_id*: int
    customers_gender*, customers_firstname*, customers_lastname*,
      customers_dob*, customers_email_address*, customers_telephone*,
      customers_fax*, customers_password*, customers_newsletter*,
      customers_activation_code*, customers_active_status*,
      customers_notes*: string
    address*: OscommerceUserAddress
  OsCommerceUserAddress* = object
    address_book_id*, customers_id*, entry_country_id*, entry_zone_id*: int
    entry_gender*, entry_company*, entry_firstname*, entry_lastname*,
        entry_street_address*, entry_suburb*, entry_postcode*, entry_city*,
        entry_state*, entry_street_address2*, entry_street_address3*: string

  WoocommerceUser* = object
    email*: string
    first_name*, last_name*, username*: string
    billing*: WoocommerceUserBilling
    shipping*: WoocommerceUserShipping
  WoocommerceUserBilling* = object
    first_name*, last_name*, company*, address_1*, address_2*, city*, state*,
      postcode*, country*, email*, phone*: string
  WoocommerceUserShipping* = object
    first_name*, last_name*, company*, address_1*, address_2*, city*, state*,
      postcode*, country*: string


src/oscommerce2woocommerce/parser.nim
#[
  Created at: 08/31/2021 21:51:13 Tuesday
  Modified at: 09/01/2021 09:38:54 PM Wednesday

        Copyright (C) 2021 Thiago Navarro
  See file "license" for details about copyright
]#

from std/strutils import find
from std/strformat import fmt
from std/tables import toTable, `[]`

import oscommerce2woocommerce/base
export base

const countryCode = {
  30: "BR",
  31: "US"
}.toTable

func toWoo*(user: OsCommerceUser): WooCommerceUser =
  result.first_name = user.customers_firstname
  result.last_name = user.customers_lastname
  result.email = user.customers_email_address
  block email:
    let atSymbol = result.email.find "@"
    if atSymbol > 0:
      result.username = result.email[0..<atSymbol]
  block billing:
    result.billing.first_name = result.first_name
    result.billing.last_name = result.last_name
    result.billing.email = result.email
    result.billing.phone = user.customers_telephone
    result.billing.address_1 = user.address.entry_street_address
    result.billing.city = user.address.entry_city
    result.billing.postcode = user.address.entry_postcode
    result.billing.country = countryCode[user.address.entry_country_id]
  block shipping:
    result.shipping.first_name = result.billing.first_name
    result.shipping.last_name = result.billing.last_name
    result.shipping.address_1 = result.billing.address_1
    result.shipping.city = result.billing.city
    result.shipping.postcode = result.billing.postcode
    result.shipping.country = result.billing.country


src/oscommerce2woocommerce.nim
#[
  Created at: 08/31/2021 21:50:31 Tuesday
  Modified at: 09/23/2021 07:19:29 PM Thursday

        Copyright (C) 2021 Thiago Navarro
  See file "license" for details about copyright
]#

import oscommerce2woocommerce/parser
export parser

{.experimental: "codeReordering".}

when isMainModule:
  from pkg/jsony import fromJson, toJson
  from std/os import fileExists

  from std/httpclient import newHttpClient, post, close, newHttpHeaders, HttpClient
  import std/httpclient
  from std/strformat import fmt
  from std/base64 import nil

  proc main(
    customersJson, addressBookJson: string;
    output = "";
    wooUrl = "";
    wooApiKey = "";
    wooApiSecret = ""
  ) =
    if not customersJson.fileExists:
      quit "The customers json file not exists."
    if not addressBookJson.fileExists:
      quit "The address_book json file not exists."
    try:
      let addresses = addressBookJson.readFile.fromJson seq[OsCommerceUserAddress]
      var
        users = customersJson.readFile.fromJson seq[OsCommerceUser]
        wooUsers: seq[WoocommerceUser]
        useApi = false
      if wooUrl.len > 0 and wooApiKey.len > 0 and wooApiSecret.len > 0:
        useApi = true
      var client: HttpClient
      if useApi:
        client = newHttpClient()
        client.headers = newHttpHeaders({
          "Content-type": "application/json",
          "Authorization": "Basic " & base64.encode fmt"{wooApiKey}:{wooApiSecret}"
        })
        defer: client.close()
      for i in 0..<users.len:
        users[i].address = addresses[i]
        let wooUser = users[i].toWoo()
        if output.len > 0:
          wooUsers.add wooUser
        if useApi:
          echo client.post(wooUrl & "/wp-json/wc/v3/customers", body = wooUser.toJson).code
          # echo "Adding user " & $users[i].customers_id
      if output.len > 0:
        writeFile output, wooUsers.toJson
    except:
      echo "Error: " & getCurrentExceptionMsg()

  import pkg/cligen
  dispatch main

#[
fetch('https://example.com/wp-json/wc/v3/customers', {
  method: 'POST',
  headers: {
    'Content-Type': 'application/json',
    'Authorization': 'Basic ' + btoa('consumer_key:consumer_secret')
  },
  body: JSON.stringify({ "email": "john.doe@example.com", "first_name": "John", "last_name": "Doe", "username": "john.doe", "billing": { "first_name": "John", "last_name": "Doe", "company": "", "address_1": "969 Market", "address_2": "", "city": "San Francisco", "state": "CA", "postcode": "94103", "country": "US", "email": "john.doe@example.com", "phone": "(555) 555-5555" }, "shipping": { "first_name": "John", "last_name": "Doe", "company": "", "address_1": "969 Market", "address_2": "", "city": "San Francisco", "state": "CA", "postcode": "94103", "country": "US" } })
});
]#
